Add support for running commands on key press

miedzinski opened the pull request:
Based on option `command` in key binding section in config, e.g.

```
  - { key: N,        mods: Control|Shift,    command: alacritty          }
  # or
  - {
      key: N,
      mods: Control|Shift,
      command: {
        program: "alacritty",
        args: ["-e", "vttest"],
  }}
```

specified command will be run in the background on key press. Alacritty doesn't wait for its result nor block IO.

Fixes #334.

jwilm commented on src/config.rs:
```diff
@@ -378,6 +378,111 @@ impl de::Deserialize for ActionWrapper {
+    args: Vec<String>,
+}
+
+impl de::Deserialize for CommandWrapper {
```
So, this whole block can be implemented automatically in this case. Just need to put a `#[derive(Deserialize)]` above `struct CommandWrapper`.

jwilm commented on src/config.rs:
```diff
@@ -378,6 +378,111 @@ impl de::Deserialize for ActionWrapper {
     }
 }
 
+struct CommandWrapper {
```
`#[derive(Deserialize)]`

jwilm commented on src/config.rs:
```diff
@@ -378,6 +378,111 @@ impl de::Deserialize for ActionWrapper {
 
+struct CommandWrapper {
+    program: String,
+    args: Vec<String>,
```
This should probably be optional, so add a `#[serde(default)]` (the default is an empty vec).

jwilm commented on src/input.rs:
```diff
@@ -174,6 +178,17 @@ impl Action {
                     });
             },
+            Action::Command(ref program, ref args) => {
+                trace!("running command: {} {}", program, args.join(" "));
```
Rather than use `args.join(" ")` here, maybe we can just pass `args` instead and use the debug formatter `{:?}` instead of `{}`.

jwilm commented on src/input.rs:
```diff
@@ -174,6 +178,17 @@ impl Action {
+                        debug!("spawned new proc with pid: {}", child.id());
+                    },
+                    Err(err) => {
+                        err_println!("couldn't run command: {}", err);
```
How do you feel about making this a `warn!` or `error!` instead?

miedzinski replied:
I swear it was `error!` until the very last second. :) Reverted.

miedzinski commented:
I have removed pretty much all deser logic and replaced it with `derive`s. The thing is there was implemented something similiar (just not generic) to https://serde.rs/string-or-struct.html, but serde doesn't provide this out of the box. I've managed to achieve the same using serde's untagged enum representation. Is it okay?

jwilm commented:
Thanks for the fixes and sorry for the delay on getting back to it. Looks good to me!
